### ISSUE TITLE
Composer Fees

### ISSUE THREAD

**Issue by aszepieniec** [Neptune-Crypto/neptune-core#224]
The two steps in the two-step mining process are *composing*, which involves selecting transactions from the mempool, merging them, and producing the block proof, and *guessing*, which involves modifying the nonce field in the header until the block hash satisfies the difficulty criterion. Two-step mining enables the composer and guesser to be separate entities that do not need to trust each other, even for the duration of one block.

Right now, the composer can send any portion of the block subsidy to the guesser. He just generates a transaction with coinbase set to the block subsidy, and with a fraction of that sum allocated to the transaction fee. The guesser automatically receives that allocated fraction when someone else builds on top of the block that the guesser contributed the proof-of-work to. Conversely, the composer can claim all the block subsidy for himself if he chooses. The distribution of the mining reward happens at his discretion, informed by the market situation.

However, right now there is no option for the block composer to collect part of the *transaction fee*, and as a result, all of it goes to the guesser. This is regrettable because it means that composers working in isolation have no incentive to include high-fee transactions. Worse still: the way you mitigate the cost of proving is by collecting transaction fees, and the only way for provers to recuperate their expenses is to participate in a mining pool where they can collect a portion of the mining fees. In other words, there is an incentive to join or run a mining pool that's separate from the desire to reduce the variance on income. This dynamic has a tendency to centralize over time.

*Composer fees* address this issue. These are UTXOs that are automatically generated and added to the mutator set, alongside guesser fee UTXOs. In the current system, the entire block transaction fee goes to the guesser. With this feature in place, the composer can claim up to all of the transaction fee, and the remainder goes to the guesser.

To make this work, the following changes must be applied:
 - the block body must contain fields for an amount and a hash image
 - these fields define composer fee UTXOs in the same way that the nonce and transaction kernel define guesser fee UTXOs
 - whenever a mutator set is updated, the mutator set update must include addition records generated from the composer fee UTXOs, just as it already includes addition records generated from the guesser fee UTXOs
 - the composer must be able to collect the fees if a guesser successfully guessed a winning nonce on the composer's block proposal

**Comment by dan-da** [comment]
this has been implemented already correct?   can it be closed?

**Comment by aszepieniec** [comment]
Yes, ages ago. I can't even determine in which commit -- there have been too many in the mean time.
